lib/types.nix

```diff
@@ -1463,8 +1561,33 @@ let
         Fixing is not trivial, we appreciate any help!
         :::
       */
-      addCheck = elemType: check: elemType // { check = x: elemType.check x && check x; };
-
+      addCheck =
+        elemType: check:
+        if elemType.merge ? v2 then
+          elemType
+          // {
+            check = x: elemType.check x && check x;
+            merge = {
+              __functor =
+                self: loc: defs:
+                (self.v2 { inherit loc defs; }).value;
```

Comment:
Will v2 types have to support v1 forever? I expect there's a lot of code out there that assumes `merge` is a function that can be called; especially out-of-tree code. Removing the functor could break a lot of assumptions, even if it was done a long way off in the future.

I don't see this as a problem, per se. Especially if test cases cover both v1 and v2 usage, for all types that implement v2.